### PR TITLE
Add metadata field to benchmark results + use Pydantic

### PR DESCRIPTION
**Pull request by tonywu71**
## Description

Add metadata field to benchmark results + use Pydantic.

**Note:** Once merged, [this PR on vidore-leaderboard](https://huggingface.co/spaces/vidore/vidore-leaderboard/discussions/2) should be reviewed and merged too.

### Added

- Add Pydantic models for storing the ViDoRe benchmark results and metadata (includes package hash)

### Example of a result JSON output

**Input command:**

```bash
vidore-benchmark evaluate-retriever \
    --model-class dummy_retriever \
    --dataset-name vidore/syntheticDocQA_dummy \
    --split test
```

`dummy_retriever_all_metrics.json`

```json
{
    "metadata": {
        "timestamp": "2024-12-01T22:17:31.271300",
        "vidore_benchmark_version": "4.0.3.dev7+g462dc4f.d20241102"
    },
    "metrics": {
        "vidore/syntheticDocQA_dummy": {
            "ndcg_at_1": 0.5,
            "ndcg_at_3": 0.5,
            "ndcg_at_5": 0.71534,
            "ndcg_at_10": 0.71534,
            "ndcg_at_20": 0.71534,
            "ndcg_at_50": 0.71534,
            "ndcg_at_100": 0.71534,
            "map_at_1": 0.5,
            "map_at_3": 0.5,
            "map_at_5": 0.625,
            "map_at_10": 0.625,
            "map_at_20": 0.625,
            "map_at_50": 0.625,
            "map_at_100": 0.625,
            "recall_at_1": 0.5,
            "recall_at_3": 0.5,
            "recall_at_5": 1.0,
            "recall_at_10": 1.0,
            "recall_at_20": 1.0,
            "recall_at_50": 1.0,
            "recall_at_100": 1.0,
            "precision_at_1": 0.5,
            "precision_at_3": 0.16667,
            "precision_at_5": 0.2,
            "precision_at_10": 0.1,
            "precision_at_20": 0.05,
            "precision_at_50": 0.02,
            "precision_at_100": 0.01,
            "mrr_at_1": 0.5,
            "mrr_at_3": 0.5,
            "mrr_at_5": 0.625,
            "mrr_at_10": 0.625,
            "mrr_at_20": 0.625,
            "mrr_at_50": 0.625,
            "mrr_at_100": 0.625,
            "naucs_at_1_max": 1.0,
            "naucs_at_1_std": -1.0,
            "naucs_at_1_diff1": 1.0,
            "naucs_at_3_max": 1.0,
            "naucs_at_3_std": -1.0000000000000002,
            "naucs_at_3_diff1": 1.0,
            "naucs_at_5_max": null,
            "naucs_at_5_std": null,
            "naucs_at_5_diff1": null,
            "naucs_at_10_max": null,
            "naucs_at_10_std": null,
            "naucs_at_10_diff1": null,
            "naucs_at_20_max": null,
            "naucs_at_20_std": null,
            "naucs_at_20_diff1": null,
            "naucs_at_50_max": 1.0,
            "naucs_at_50_std": 1.0,
            "naucs_at_50_diff1": 1.0,
            "naucs_at_100_max": 1.0,
            "naucs_at_100_std": 1.0,
            "naucs_at_100_diff1": 1.0
        }
    }
}
```

### Check metric consistency wrt HEAD of `main` (f1f40c1e610fbe41d348d94c8faad6ce0bdb2131)

```bash
vidore-benchmark evaluate-retriever \
    --model-class colpali \
    --model-name vidore/colpali-v1.2 \
    --dataset-name vidore/shiftproject_test \
    --split test
```

**Results from HEAD:**

```json
{
    "vidore/shiftproject_test": {
        "ndcg_at_1": 0.62,
        "ndcg_at_3": 0.75726,
        "ndcg_at_5": 0.78222,
        "ndcg_at_10": 0.79496,
        "ndcg_at_20": 0.80225,
        "ndcg_at_50": 0.80407,
        "ndcg_at_100": 0.80407,
        "map_at_1": 0.62,
        "map_at_3": 0.725,
        "map_at_5": 0.739,
        "map_at_10": 0.74414,
        "map_at_20": 0.74599,
        "map_at_50": 0.74622,
        "map_at_100": 0.74622,
        "recall_at_1": 0.62,
        "recall_at_3": 0.85,
        "recall_at_5": 0.91,
        "recall_at_10": 0.95,
        "recall_at_20": 0.98,
        "recall_at_50": 0.99,
        "recall_at_100": 0.99,
        "precision_at_1": 0.62,
        "precision_at_3": 0.28333,
        "precision_at_5": 0.182,
        "precision_at_10": 0.095,
        "precision_at_20": 0.049,
        "precision_at_50": 0.0198,
        "precision_at_100": 0.0099,
        "mrr_at_1": 0.65,
        "mrr_at_3": 0.7366666666666667,
        "mrr_at_5": 0.7531666666666667,
        "mrr_at_10": 0.7600396825396826,
        "mrr_at_20": 0.7613095238095237,
        "mrr_at_50": 0.7615367965367965,
        "mrr_at_100": 0.7615367965367965,
        "naucs_at_1_max": 0.011383564594037348,
        "naucs_at_1_std": -0.3069255145678863,
        "naucs_at_1_diff1": 0.6713226471402637,
        "naucs_at_3_max": -0.09118125610152811,
        "naucs_at_3_std": -0.45034168564920113,
        "naucs_at_3_diff1": 0.4871461112918982,
        "naucs_at_5_max": -0.42047930283224205,
        "naucs_at_5_std": -0.9665421724245231,
        "naucs_at_5_diff1": 0.5389563232700478,
        "naucs_at_10_max": -0.4455648926237065,
        "naucs_at_10_std": -1.190943043884201,
        "naucs_at_10_diff1": 0.7343604108310002,
        "naucs_at_20_max": -0.6909430438842098,
        "naucs_at_20_std": -1.1517273576096962,
        "naucs_at_20_diff1": 0.861111111111116,
        "naucs_at_50_max": 0.35807656395892007,
        "naucs_at_50_std": -0.5634920634920583,
        "naucs_at_50_diff1": 0.7222222222222041,
        "naucs_at_100_max": 0.35807656395892007,
        "naucs_at_100_std": -0.5634920634920583,
        "naucs_at_100_diff1": 0.7222222222222041
    }
}
```

**Results with this PR branch:**

```json
{
    "metadata": {
        "timestamp": "2024-12-02T15:57:02.316118",
        "vidore_benchmark_version": "4.0.3.dev56+g6611681"
    },
    "metrics": {
        "vidore/shiftproject_test": {
            "ndcg_at_1": 0.62,
            "ndcg_at_3": 0.75726,
            "ndcg_at_5": 0.78222,
            "ndcg_at_10": 0.79496,
            "ndcg_at_20": 0.80225,
            "ndcg_at_50": 0.80407,
            "ndcg_at_100": 0.80407,
            "map_at_1": 0.62,
            "map_at_3": 0.725,
            "map_at_5": 0.739,
            "map_at_10": 0.74414,
            "map_at_20": 0.74599,
            "map_at_50": 0.74622,
            "map_at_100": 0.74622,
            "recall_at_1": 0.62,
            "recall_at_3": 0.85,
            "recall_at_5": 0.91,
            "recall_at_10": 0.95,
            "recall_at_20": 0.98,
            "recall_at_50": 0.99,
            "recall_at_100": 0.99,
            "precision_at_1": 0.62,
            "precision_at_3": 0.28333,
            "precision_at_5": 0.182,
            "precision_at_10": 0.095,
            "precision_at_20": 0.049,
            "precision_at_50": 0.0198,
            "precision_at_100": 0.0099,
            "mrr_at_1": 0.65,
            "mrr_at_3": 0.7366666666666667,
            "mrr_at_5": 0.7531666666666667,
            "mrr_at_10": 0.7600396825396826,
            "mrr_at_20": 0.7613095238095237,
            "mrr_at_50": 0.7615367965367965,
            "mrr_at_100": 0.7615367965367965,
            "naucs_at_1_max": 0.011383564594037348,
            "naucs_at_1_std": -0.3069255145678863,
            "naucs_at_1_diff1": 0.6713226471402637,
            "naucs_at_3_max": -0.09118125610152811,
            "naucs_at_3_std": -0.45034168564920113,
            "naucs_at_3_diff1": 0.4871461112918982,
            "naucs_at_5_max": -0.42047930283224205,
            "naucs_at_5_std": -0.9665421724245231,
            "naucs_at_5_diff1": 0.5389563232700478,
            "naucs_at_10_max": -0.4455648926237065,
            "naucs_at_10_std": -1.190943043884201,
            "naucs_at_10_diff1": 0.7343604108310002,
            "naucs_at_20_max": -0.6909430438842098,
            "naucs_at_20_std": -1.1517273576096962,
            "naucs_at_20_diff1": 0.861111111111116,
            "naucs_at_50_max": 0.35807656395892007,
            "naucs_at_50_std": -0.5634920634920583,
            "naucs_at_50_diff1": 0.7222222222222041,
            "naucs_at_100_max": 0.35807656395892007,
            "naucs_at_100_std": -0.5634920634920583,
            "naucs_at_100_diff1": 0.7222222222222041
        }
    }
}
```

No diff in the metrics!

## Next steps: ViDoRe Leaderboard

Once merged, the following PR for the `vidore-leaderboard` space should be merged as well: https://huggingface.co/spaces/vidore/vidore-leaderboard/discussions/2.

**Test:**

Screenshot 2024-12-01 at 22 42 50

### CODE REVIEW

File: src/vidore_benchmark/evaluation/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+from datetime import datetime
+from typing import Dict, List, Optional
+
+from pydantic import BaseModel
+
+
+class MetricsModel(BaseModel):
```
Comment (ManuelFay) [inline src/vidore_benchmark/evaluation/interfaces.py:7]: Problem with this is that  might break in fure mteb releases ?

File: src/vidore_benchmark/retrievers/dse_qwen2_retriever.py [hunk 2]
```diff
@@ -173,10 +173,13 @@ def get_scores(
         passage_embeddings: Union[torch.Tensor, List[torch.Tensor]],
         batch_size: Optional[int] = None,
     ) -> torch.Tensor:
+        """
+        Dot-product similarity between queries and passages.
+        """
         if isinstance(query_embeddings, list):
-            query_embeddings = torch.stack(query_embeddings).to(self.device)
+            query_embeddings = torch.stack(query_embeddings)
```
Comment (ManuelFay) [inline src/vidore_benchmark/retrievers/dse_qwen2_retriever.py:180]: I'm guessing the to self.device comes layer ?

File: src/vidore_benchmark/retrievers/dummy_retriever.py [hunk 3]
```diff
@@ -14,34 +13,45 @@
...
+        self.device = device
 
     @property
     def use_visual_embedding(self) -> bool:
         return True
 
-    def forward_queries(self, queries: List[str], batch_size: int, **kwargs) -> List[Tensor]:
-        return [torch.randn(batch_size, self.emb_dim_query) for _ in range(math.ceil(len(queries) / batch_size))]
+    def forward_queries(self, queries: List[str], batch_size: int, **kwargs) -> Tensor:
+        return torch.randn(len(queries), self.emb_dim_query).to(self.device)
```
Comment (ManuelFay) [inline src/vidore_benchmark/retrievers/dummy_retriever.py:36]: why the torch randn ? was it just to mock ?

**Review by ManuelFay — COMMENTED** [review]
One thing yuou removed is the .to(device) for the embeddings - but you did not put it back in each model's scoring func ? this might make things slow ?

File: src/vidore_benchmark/retrievers/dummy_retriever.py (hunk 3 above)
Comment (tonywu71) [reply src/vidore_benchmark/retrievers/dummy_retriever.py:36]: `torch.randn` is used as part of the the `dummy_retriever` model class, and indeed I'm using it in my `test_evaluate_retriever.py` which tests the `evaluate-retriever` CLI command on the [`vidore/syntheticDocQA_dummy`](https://huggingface.co/datasets/vidore/syntheticDocQA_dummy) dummy dataset. While I could have mocked another of our retrievers, I decided having a dummy retriever was simpler.

The exact command that is run in this test is:

```bash
vidore-benchmark evaluate-retriever \
    --model-class dummy_retriever \
    --dataset-name vidore/syntheticDocQA_dummy \
    --split test \
    --batch-query 2 \
    --batch-passage 2 \
    --batch-score 2
```

File: src/vidore_benchmark/retrievers/dse_qwen2_retriever.py (hunk 2 above)
Comment (tonywu71) [reply src/vidore_benchmark/retrievers/dse_qwen2_retriever.py:180]: I have removed the `.to(self.device)` to make the `DSEQwen2Retriever` ISO with the others and because I think it's better to let the user the responsibility of sending the tensors fed to the `get_scores` method.

In particular, keeping `.to(self.device)` made my VM crashed for large datasets. For this reason, I have moved the computed embeddings to `cpu` in both the `forward_passages` and `forward_queries` method.

File: src/vidore_benchmark/evaluation/interfaces.py (hunk 1 above)
Comment (tonywu71) [reply src/vidore_benchmark/evaluation/interfaces.py:7]: You're right, plus the class is a bit bloated tbh... I'll simply make `MetricsModel` a `Dict[str, Optional[float]]`!
